File: python/Q110_balanced_binary_tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def is_balanced(self, root: TreeNode | None) -> tuple[int, bool]:
        if not root:
            return 0, True

        left_depth, left_balanced = self.is_balanced(root.left)
        if not left_balanced:
            return -1, False
        right_depth, right_balanced = self.is_balanced(root.right)
        if not right_balanced:
            return -1, False

        depth: int = left_depth if left_depth > right_depth else right_depth
        depth += 1

        return depth, abs(left_depth - right_depth) < 2

    def isBalanced(self, root: TreeNode | None) -> bool:
        return self.is_balanced(root)[1]


# @lc code=end

File: rust/src/leetcode/Q110_balanced_binary_tree.rs
/*
 * @lc app=leetcode id=110 lang=rust
 *
 * [110] Balanced Binary Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::cmp::max;
use std::rc::Rc;

impl Solution {
    fn is_balanced_with_depth(root: Option<Rc<RefCell<TreeNode>>>) -> (bool, i32) {
        if let Some(node) = root {
            let (left_balanced, left_depth) =
                Self::is_balanced_with_depth(node.borrow_mut().left.clone());
            let (right_balanced, right_depth) =
                Self::is_balanced_with_depth(node.borrow_mut().right.clone());

            let self_balanced: bool =
                left_balanced && right_balanced && (left_depth - right_depth).abs() <= 1;

            (self_balanced, max(left_depth, right_depth) + 1)
        } else {
            (true, 0)
        }
    }

    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::is_balanced_with_depth(root).0
    }
}
// @lc code=end
